### PR TITLE
Ensure unique attributes on telemetry

### LINKED ISSUE #6178: Error viewing some telemetry caused by duplicate attributes
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

We added row keys to grids in the dashboard. For properties, the name is the default key. It must be unique or `FluentDataGrid` will error. When testing I found some log entries that errored because they had duplicate attributes. e.g. two `HttpMethod` attributes:

image

### Expected Behavior

_No response_

### Steps To Reproduce

All HTTP request logs have duplicate attributes.

### Exceptions (if any)

```
fail: Aspire.Hosting.Dashboard.Microsoft.AspNetCore.Components.Server.Circuits.CircuitHost[111]
      Unhandled exception in circuit 'Qhb0JkmRqSJWXFxn_LSsozjnU9oY5ow5J9WoDDfOKvY'.
      System.InvalidOperationException: More than one sibling of component 'Microsoft.FluentUI.AspNetCore.Components.FluentDataGridRow`1[Aspire.Dashboard.Model.LogEntryPropertyViewModel]' has the same key value, 'HttpMethod'. Key values must be unique.
         at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.ThrowExceptionForDuplicateKey(Object key, RenderTreeFrame& frame)
         at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.BuildKeyToInfoLookup(DiffContext diffContext, Int32 oldStartIndex, Int32 oldEndIndexExcl, Int32 newStartIndex, Int32 newEndIndexExcl)
         at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.AppendDiffEntriesForRange(DiffContext& diffContext, Int32 oldStartIndex, Int32 oldEndIndexExcl, Int32 newStartIndex, Int32 newEndIndexExcl)
         at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.AppendDiffEntriesForFramesWithSameSequence(DiffContext& diffContext, Int32 oldFrameIndex, Int32 newFrameIndex)
         at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.AppendDiffEntriesForRange(DiffContext& diffContext, Int32 oldStartIndex, Int32 oldEndIndexExcl, Int32 newStartIndex, Int32 newEndIndexExcl)
         at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.AppendDiffEntriesForFramesWithSameSequence(DiffContext& diffContext, Int32 oldFrameIndex, Int32 newFrameIndex)
         at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.AppendDiffEntriesForRange(DiffContext& diffContext, Int32 oldStartIndex, Int32 oldEndIndexExcl, Int32 newStartIndex, Int32 newEndIndexExcl)
         at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.AppendDiffEntriesForFramesWithSameSequence(DiffContext& diffContext, Int32 oldFrameIndex, Int32 newFrameIndex)
         at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.AppendDiffEntriesForRange(DiffContext& diffContext, Int32 oldStartIndex, Int32 oldEndIndexExcl, Int32 newStartIndex, Int32 newEndIndexExcl)
         at Microsoft.AspNetCore.Components.Rendering.ComponentState.RenderIntoBatch(RenderBatchBuilder batchBuilder, RenderFragment renderFragment, Exception& renderFragmentException)
         at Microsoft.AspNetCore.Components.RenderTree.Renderer.ProcessRenderQueue()
```

### .NET Version info

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by JamesNK**
## Description

We added row keys to grids in the dashboard. For properties, the name is the default key. It must be unique or `FluentDataGrid` will error. When testing I found some log entries that errored because they had duplicate attributes. e.g. two `HttpMethod` attributes.

image

The duplicate attributes are coming are coming from OTLP (i.e. they're external data). We shouldn't be sent duplicate data, but it's not something we can control.

To fix this problem the OTLP collector services only store unique attributes. This matches behavior in [the spec around attribute collections](https://opentelemetry.io/docs/specs/otel/common/#attribute-collections). Last value wins. To help identify if this is a problem, debug logs are written if there are duplicate keys with different values.

Fixes https://github.com/dotnet/aspire/issues/6178

Should be backported to RC1 if possible.

## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
  - [ ] No
- Did you add public API?
  - [ ] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
    - Link to aspire-docs issue: 
  - [x] No

### CODE REVIEW

File: src/Aspire.Dashboard/Otlp/Model/OtlpHelpers.cs [hunk 1]
```diff
@@ -202,59 +209,126 @@ public static KeyValuePair<string, string>[] ToKeyValuePairs(this RepeatedField<
 
             var value = TruncateString(attribute.Value.GetString(), options.MaxAttributeLength);
 
-            values.Add(new KeyValuePair<string, string>(attribute.Key, value));
-
-            if (values.Count >= readLimit)
+            // If there are duplicates then last value wins.
+            var existingIndex = GetIndex(values, attribute.Key);
```
Comment (drewnoakes) [inline src/Aspire.Dashboard/Otlp/Model/OtlpHelpers.cs:213]: How large can `values` get? This looks like it's O(N^2).

File: src/Aspire.Dashboard/Otlp/Model/OtlpHelpers.cs [hunk 2]
```diff
@@ -202,59 +209,126 @@ public static KeyValuePair<string, string>[] ToKeyValuePairs(this RepeatedField<
...
+        copyCount = 0;
+        for (var i = 0; i < attributes.Count; i++)
         {
             var attribute = attributes[i];
 
             var value = TruncateString(attribute.Value.GetString(), options.MaxAttributeLength);
 
-            copiedAttributes[i + index] = new KeyValuePair<string, string>(attribute.Key, value);
+            // If there are duplicates then last value wins.
+            var existingIndex = GetIndex(copiedAttributes, attribute.Key);
```
Comment (drewnoakes) [inline src/Aspire.Dashboard/Otlp/Model/OtlpHelpers.cs:290]: Ditto, O(N^2).

File: src/Aspire.Dashboard/Otlp/Model/OtlpHelpers.cs (hunk 1 above)
Comment (JamesNK) [reply src/Aspire.Dashboard/Otlp/Model/OtlpHelpers.cs:213]: There is an upper limit of 128. In the real world there probably won't be more than 10 items.

Comment (drewnoakes) [reply src/Aspire.Dashboard/Otlp/Model/OtlpHelpers.cs:212]: Ok. I plan on grabbing some ETL traces from the dashboard under various loads at some point and looking for any perf bottlenecks. If this is a problem, it'll show up then.

**Review by drewnoakes — APPROVED** [review]

**Pull request merged**